### DISCUSSION TITLE
Cucumber report

### DISCUSSION THREAD

**Discussion by anuragh27crony** [go-bdd/gobdd#126] [Q&A]
Hi All,

kudos to the team for providing go-bdd library. We are actively using for testing purposes without missing go test features.

However, just wondering if it's possible to generate cucumber format json report. I checked the docs and also the code but couldn't find any hints.

**Comment by bkielbasa** [comment]
Hi, thanks for your question!

We don't provide this feature but I can do a small research about how it can be done. Do you use any tool that needs this kind of report?

**Reply by anuragh27crony** [reply]
Sure, the use-case is to publish human-readable reports as part of build pipeline. For other projects, we are using cucumber-reports plugin from jenkins (which expects json formatted data).

https://github.com/jenkinsci/cucumber-reports-plugin

**Reply by bkielbasa** [reply]
thanks! I'll back to you with the feedback soon :)

**Reply by bkielbasa** [reply]
From what I can see, the JSON format is called "legacy" and I don't think they will support it very heavily. But they use a concept of [messages](https://github.com/cucumber/cucumber/tree/master/messages). Those messages can be used to generate JSON files as you need https://github.com/cucumber/cucumber/tree/master/json-formatter

and I'd go this way. The question is how do we want to get this report? My first idea is to add a new method to GoBDD struct like this
```
suite := gobdd.NewSuite(t)
suite.Report("path/to/file.message")
suite.Run()
```

and you'll be able to convert it to the JSON file (or anything else) as needed. WDYT?

**Reply by anuragh27crony** [reply]
thanks for thinking it through. IMO this way (generating messages) seems more logical and in-line with official cucumber team efforts. Generating json from it is easy and straight forward (across all popular platforms).

Instead of new method, how about extending NewSuite method since it's more of configuring test suite for generating report. Of course we might end up break backward compatibility as a result.

**Reply by bkielbasa** [reply]
Could you write an issue with your proposal?

Here's my second suggestion:
```go
s := gobdd.NewSuite(t).WithReport("path/to/file.message")
```

or the `WithReport` method may accept `io.Writer`. This would require a bit more code in the test suite but would give us more flexibility.

```go
f, err := os.Create("path/to/file.message")

s := gobdd.NewSuite(t).WithReport(f)// instead of os.File it can be an AWS backet etc
```

the question is if it's needed.

Another idea is to abstract it a bit more.

```go
f, err := gobdd.NewFileReporter("path/to/file.message")

s := gobdd.NewSuite(t).WithReporter(f)
```

WDYT?

**Reply by anuragh27crony** [reply]
Sure, will do that.  
Approach@2 seems good and decent.
Approach@3 with Abstraction looks cleaner approach and is extensible for any future use cases compared to just with io.writer.

**Reply by mirogta** [reply]
Approach@3 with `gobdd.NewFileReporter(...)` 👍

**Comment by vvsandeepreddy** [comment]
Hello, 

Just checking to see if this is implemented already? Please let me know. 

Thanks!

**Reply by bkielbasa** [reply]
not yet, unfortunately. Is it important feature to you?

**Reply by vvsandeepreddy** [reply]
Yes @bkielbasa, we have migrated most of our infra test cases to GoBDD frame work and are planning to stick with this one. Right now, we are looking at the test logs for results, there are a bunch of logs since we are testing lot of components. Helps a lot if there is a reporting feature with in GoBDD like discussed above.

Do you know when this feature will be ready? Based on that, we will devise a temporary solution of our own. Please let me know. 

Thanks!

**Reply by bkielbasa** [reply]
thanks for your response. I cannot promise any deadline but I'll start working on the issue soon. I'll let you know about the progress.
